### PR TITLE
Initial implementation digit packing for mpz numeral generation

### PR DESCRIPTION
**Pull request by joe-hauns**
Initial implemenation as discussed with Nikolaj.
Have a look, and let me know what you think. 
I did not know how to generate the API logging macros `LOG_Z3_mk_mpz_numeral`. This is why the related code is commented out for now. 

Further I some functions in `mpff.h` and `mpfx.h` assumed that `digit_t == unsigned`. This is why i replaced `set` by `set_digits_unsigned` in these functions. 

If you let me know how to generate the logging macro(s) i will add that.
Further how do you do testing of your apis?

### CODE REVIEW

File: src/api/api_numeral.cpp [hunk 1]
```diff
@@ -459,33 +459,36 @@ extern "C" {
         Z3_CATCH_RETURN(nullptr);
     }
 
-#if 0
-    Z3_ast Z3_API Z3_mk_mpz_numeral(Z3_context c, bool sign, unsigned n, unsigned const nums[], Z3_sort* srt) {
-        LOG_TRY;
-        LOG_Z3_mk_mpz_numeral(c, sign, n, nums, srt);
+    Z3_ast Z3_API Z3_mk_mpz_numeral(Z3_context c, bool sign, unsigned n, unsigned const nums[], Z3_sort srt) {
+        // LOG_TRY;
+        // LOG_Z3_mk_mpz_numeral(c, sign, n, nums, srt);
```
Comment (NikolajBjorner) [inline src/api/api_numeral.cpp:464]: I can take care of these

File: src/api/api_numeral.cpp [hunk 2]
```diff
@@ -459,33 +459,36 @@ extern "C" {
...
-        rational q;
 
+        rational value(n, nums, d, dens);
         if (!sign)
-            q.neg();
+            value.neg();
 
-        Z3_CATCH_RETURN(nullptr);
+        sort* real_ty = mk_c(c)->m().mk_sort(mk_c(c)->get_arith_fid(), REAL_SORT);
+        auto* a = mk_c(c)->mk_numeral_core(value, real_ty);
```
Comment (NikolajBjorner) [inline src/api/api_numeral.cpp:488]: TBD: add a to the trail before returning.

File: src/api/z3_api.h [hunk 3]
```diff
@@ -3399,7 +3399,7 @@ extern "C" {
        future_('Z3_mk_mpz_numeral', AST, (_in(CONTEXT), _in(BOOL), _in(UINT), _in_array(2, UINT), _in(SORT)))
      */
 
-    //Z3_ast Z3_mk_mpz_numeral(Z3_context c, bool sign, unsigned n, unsigned const nums[], Z3_sort s);
+    Z3_ast Z3_mk_mpz_numeral(Z3_context c, bool sign, unsigned n, unsigned const nums[], Z3_sort s);
```
Comment (NikolajBjorner) [inline src/api/z3_api.h:3402]: Add Z3_API here

File: src/api/z3_api.h [hunk 4]
```diff
@@ -3399,7 +3399,7 @@ extern "C" {
        future_('Z3_mk_mpz_numeral', AST, (_in(CONTEXT), _in(BOOL), _in(UINT), _in_array(2, UINT), _in(SORT)))
```
Comment (NikolajBjorner) [inline src/api/z3_api.h:3399]: replace _future by def_API

File: src/api/api_numeral.cpp [hunk 5]
```diff
@@ -459,33 +459,36 @@ extern "C" {
...
-        rational z;
 
-        // todo fill in z
-        if (!z.size())
-            z.neg();
-        arith_util & a = mk_c(c)->autil();
-        auto* a = mk_c(c)->mk_numeral_core(r, a.mk_int_sort());
-        Z3_CATCH_RETURN(nullptr);
+        unsigned den[] = {1};
+        rational value(n, nums, 1, den);
```
Comment (NikolajBjorner) [inline src/api/api_numeral.cpp:468]: we have constructor that takes a single numerator too

File: src/api/api_numeral.cpp [hunk 6]
```diff
@@ -459,33 +459,36 @@ extern "C" {
...
+        if (!sign)
+            value.neg();
```
Comment (NikolajBjorner) [inline src/api/api_numeral.cpp:470]: Coersion into digit_t is TBD.
```
#if sizeof(unsigned) == sizeof(digit_t)

// no coercion

#else 
// compile assert sizeof(digit_t) == 8
buffer<unsigned> _num;
unsigned i = 0;
for (; i + 1 < n; i += 2) 
buffer.push_back((digit_t)num[i] + (digit_t))num[i+1] << 32ull);
if (i < n) 
   buffer.push_back(num[i]);

rational value(buffer.size(), buffer.data());
#endif
```

File: src/api/api_numeral.cpp (hunk 1 above)
Comment (joe-hauns) [reply src/api/api_numeral.cpp:464]: great, thanks!

File: src/api/api_numeral.cpp (hunk 6 above)
Comment (joe-hauns) [reply src/api/api_numeral.cpp:470]: The coersion already happens now, but in the lowest level (in `mpz.h` line 438ff).

Comment (NikolajBjorner) [reply src/api/api_numeral.cpp:469]: It is not yet part of your pull request.

File: src/util/mpz.cpp [hunk 7]
```diff
@@ -435,6 +435,57 @@ void mpz_manager<SYNCH>::set(mpz & a, char const * val) {
...
+
+  if (std::is_same<digit_t, unsigned>::value) {
+    set_digits(target, sz, digits);
+
+  } else if (little_endian && sz % us_per_digit == 0) {
+    set_digits(target, new_sz, (digit_t const*) digits);
+
+  } else {
+    // we need to allocate a new buffer
+    auto new_digits = new digit_t[new_sz];
```
Comment (nunoplopes) [inline src/util/mpz.cpp:457]: use unique_ptr please:
`auto new_digits = std::make_unique<digit_t[]>(new_sz);`

Then the delete below can go away.

**Comment by NikolajBjorner** [comment]
@joe-hauns - do you plan to finish this pull request?

**Pull request closed without merging**